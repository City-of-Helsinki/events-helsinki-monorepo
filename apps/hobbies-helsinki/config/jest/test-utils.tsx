import type { ApolloCache, InMemoryCache } from '@apollo/client/cache';
import type { MockedResponse } from '@apollo/client/testing';
import type { RenderResult } from '@testing-library/react';
import { act, fireEvent, render } from '@testing-library/react';
import type { NextRouter } from 'next/router';
import Router from 'next/router';
import type { ReactElement } from 'react';
import React from 'react';
import wait from 'waait';

import { hobbiesApolloClient } from '../../src/domain/clients/hobbiesApolloClient';
import TestProviders from './TestProviders';

type CustomRender = {
  (
    ui: React.ReactElement,
    options?: {
      mocks?: MockedResponse[];
      cache?: ApolloCache<Record<string, unknown>> | InMemoryCache;
      routes?: string | string[];
    }
  ): CustomRenderResult;
};

export type CustomRenderResult = RenderResult & { router: NextRouter };

export const arrowUpKeyPressHelper = (): boolean =>
  fireEvent.keyDown(document, { code: 38, key: 'ArrowUp' });

export const arrowDownKeyPressHelper = (): boolean =>
  fireEvent.keyDown(document, { code: 40, key: 'ArrowDown' });

export const enterKeyPressHelper = (): boolean =>
  fireEvent.keyDown(document, { code: 13, key: 'Enter' });

export const escKeyPressHelper = (): boolean =>
  fireEvent.keyDown(document, { code: 27, key: 'Escape' });

export const tabKeyPressHelper = (): boolean =>
  fireEvent.keyDown(document, { code: 9, key: 'Tab' });

const customRender: CustomRender = (
  ui: ReactElement,
  { mocks = [], cache = hobbiesApolloClient.cache, routes = [] } = {}
) => {
  if (routes) {
    if (!Array.isArray(routes)) {
      routes = [routes];
    }
    routes.forEach((route) => Router.push(route));
  }

  // set locales so that the routing tests for default locale worked
  Router.locales = ['fi', 'en', 'sv'];

  const renderResult = render(ui, {
    wrapper: ({ children }) => (
      <TestProviders mocks={mocks} router={Router} cache={cache}>
        {children}
      </TestProviders>
    ),
  });
  return {
    ...renderResult,
    router: Router,
  };
};

// eslint-disable-next-line testing-library/no-unnecessary-act
const actWait = (amount?: number): Promise<void> => act(() => wait(amount));

// eslint-disable-next-line import/export
export { actWait, customRender as render };

// re-export everything
// eslint-disable-next-line import/export
export * from '@testing-library/react';
export { render as defaultRender } from '@testing-library/react';
export { default as userEvent } from '@testing-library/user-event';
