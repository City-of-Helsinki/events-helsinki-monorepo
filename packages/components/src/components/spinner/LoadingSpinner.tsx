import classNames from 'classnames';
import type { LoadingSpinnerCustomTheme, LoadingSpinnerProps } from 'hds-react';
import { LoadingSpinner as HDSLoadingSpinner } from 'hds-react';
import type { FunctionComponent } from 'react';
import React from 'react';

import useCommonTranslation from '../../hooks/useCommonTranslation';
import styles from './loadingSpinner.module.scss';

type Props = {
  hasPadding?: boolean;
  isLoading: boolean;
  children?: React.ReactNode;
} & Omit<LoadingSpinnerProps, 'loadingFinishedText' | 'loadingText'>;

const defaultTheme: LoadingSpinnerCustomTheme = {
  '--spinner-color-stage1': 'var(--color-coat-of-arms)',
  '--spinner-color-stage2': 'var(--color-copper)',
  '--spinner-color-stage3': 'var(--color-suomenlinna)',
};

const LoadingSpinner: FunctionComponent<Props> = ({
  hasPadding = true,
  isLoading,
  children,
  className,
  multicolor = true,
  theme = defaultTheme,
  ...rest
}) => {
  const { t } = useCommonTranslation();
  const loadingFinishedText = t('common:loadingSpinner.loadingFinishedText');
  const loadingText = t('common:loadingSpinner.loadingText');

  return (
    <>
      {isLoading ? (
        <div
          className={classNames(styles.spinnerWrapper, className, 'spinner', {
            [styles.hasPadding]: hasPadding,
          })}
          data-testid="loading-spinner"
        >
          <HDSLoadingSpinner
            {...rest}
            multicolor={multicolor}
            loadingFinishedText={loadingFinishedText}
            loadingText={loadingText}
            theme={theme}
          />
        </div>
      ) : (
        children
      )}
    </>
  );
};

export default LoadingSpinner;
