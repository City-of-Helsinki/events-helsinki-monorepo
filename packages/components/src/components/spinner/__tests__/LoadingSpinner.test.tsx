import React from 'react';
import { render, screen } from '@/test-utils';

import LoadingSpinner from '../LoadingSpinner';

it('matches snapshot', () => {
  const { container } = render(<LoadingSpinner isLoading={true} />);
  expect(container.firstChild).toMatchSnapshot();
});

it('render spinner if isLoading is true', () => {
  render(<LoadingSpinner isLoading={true} />);
  expect(screen.getByTestId('loading-spinner')).toBeInTheDocument();
});

it('render child component if isLoading is false', () => {
  const { container } = render(
    <LoadingSpinner isLoading={false}>
      <div className="component"></div>
    </LoadingSpinner>
  );
  expect((container.firstChild as HTMLElement).classList).toContain(
    'component'
  );
});
