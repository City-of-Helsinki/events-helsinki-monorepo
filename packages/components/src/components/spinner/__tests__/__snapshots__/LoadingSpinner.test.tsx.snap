// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`matches snapshot 1`] = `
<div
  class="_spinnerWrapper_d3ce5a spinner _hasPadding_d3ce5a"
  data-testid="loading-spinner"
>
  <div
    class="LoadingSpinner-module_loadingSpinner__2xHT_ LoadingSpinner-module_multicolor__1qbEF custom-theme-1"
  >
    <div />
    <div />
    <div />
  </div>
</div>
`;
